### ISSUE TITLE
Build DataFrames from CUDA array libraries

### ISSUE THREAD

**Issue by mrocklin** [rapidsai/cudf#529]
## Feature request

I would like to be able to construct a cudf DataFrame from a NumPy-like GPU array, perhaps coming from Numba, CuPy, PyTorch, or TensorFlow.  What is the right way to support this?

Alternatively, if I want to move a cudf Series to a CuPy array, or work with it from Numba, how can one do this?

One approach was defined and taken in https://github.com/cupy/cupy/pull/1144 where @seibert added a `__cuda_array_interface__` protocol to cupy, enabling Numba-defined GPU functions to work on that data directly.  I'm curious what it would take to support the same within cudf (Numba working on cudf objects), and also to allow cudf and cupy to share data between each other.  

I suspect that there are challenges here, but thought I'd go ahead and ask.

**Comment by mrocklin** [comment]
Actual docs here: https://numba.pydata.org/numba-doc/dev/cuda/cuda_array_interface.html

**Comment by mrocklin** [comment]
My objective would be for the following to work:

```python
import cudf, cupy

x = cupy.random.random(10000)
df = cudf.DataFrame({'x': x})
```

**Comment by kkraus14** [comment]
@mrocklin this definitely doable. Would you expect the behavior to be pass by reference or pass by copy?


With Pandas + Numpy the behavior is inconsistent:

In constructing a Pandas Series from a Numpy array it is pass by reference:

```python
x = np.array([1,2,3,4,5])
y = pd.Series(x)
y[:] = 10
print(x)
```

```
[10 10 10 10 10]
```

In constructing a Pandas DataFrame from Numpy Array(s) it is pass by copy if constructed via a dict:

```python
x = np.array([1,2,3,4,5])
pdf = pd.DataFrame({'y': x})
pdf['y'][:] = 10
print(x)
```

```
[1 2 3 4 5]
```

In constructing a Pandas DataFrame from a Numpy Array it is pass by reference if constructed directly:

```python
x = np.array([1,2,3,4,5])
pdf = pd.DataFrame(x)
pdf[0][:] = 10
print(x)
```

```
[10 10 10 10 10]
```

**Comment by mrocklin** [comment]
I'm personally comfortable with pass-by-reference whenever possible.  I'm a bit curious about who owns the reference though.  What happens in the following case?

```python
x = cupy.random.random(...)
s = cudf.Series(x)
del x
```

Ideally `s` would still point to valid data, but I can imagine that the memory manager within cupy may not be aware of this.  Does `s` hold a reference to `x`?

**Comment by mrocklin** [comment]
Given the complexity behind different memory managers, I would probably start by copying just for ease.

**Comment by kkraus14** [comment]
Being tackled in #822

**Comment by kkraus14** [comment]
Resolved by #822.
